core/src/main/java/org/apache/calcite/rel/rules/JoinConditionOrExpansionRule.java

```diff
@@ -140,31 +141,55 @@ private List<RexNode> splitCond(Join join) {
   }
 
   private boolean isValidCond(RexNode node, int leftFieldCount) {
-    if (!(node instanceof RexCall)) {
-      return false;
-    }
+    boolean hasJoinKeyCond = false;
+    List<RexNode> conds = RelOptUtil.conjunctions(node);
+    for (RexNode cond : conds) {
```

Comment:
Frankly, I don't see where "column and constants" is handled
These checks are fine, but I would have expected to also see a check for one-sidedness.
This can be done with a visitor which checks that the RexInputRefs that appear in the expression are not simultaneously from the left and right sides (one side or no side (i.e. constants) are fine).